File: app/api/routes/ml.py
from __future__ import annotations

import uuid

import httpx
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlmodel import select

from app.api.deps import CurrentUser, SessionDep, get_current_active_superuser
from app.core.config import settings
from app.models import MLModelRegistry, MLOfflineRiskPrediction, MLTonerPrediction
from app.schemas import (
    Message,
    MLModelsStatusPublic,
    MLOfflineRiskPredictionsPublic,
    MLTonerPredictionsPublic,
)

router = APIRouter(tags=["ml"])


@router.get("/predictions/toner", response_model=MLTonerPredictionsPublic)
def read_toner_predictions(
    session: SessionDep,
    current_user: CurrentUser,
    printer_id: uuid.UUID | None = Query(default=None),
    limit: int = Query(default=200, ge=1, le=500),
) -> MLTonerPredictionsPublic:
    del current_user
    statement = select(MLTonerPrediction).order_by(MLTonerPrediction.created_at.desc())
    if printer_id is not None:
        statement = statement.where(MLTonerPrediction.printer_id == printer_id)
    rows = session.exec(statement.limit(limit)).all()
    return MLTonerPredictionsPublic(data=rows, count=len(rows))


@router.get("/predictions/offline-risk", response_model=MLOfflineRiskPredictionsPublic)
def read_offline_risk_predictions(
    session: SessionDep,
    current_user: CurrentUser,
    device_kind: str | None = Query(default=None),
    limit: int = Query(default=300, ge=1, le=1000),
) -> MLOfflineRiskPredictionsPublic:
    del current_user
    statement = select(MLOfflineRiskPrediction).order_by(MLOfflineRiskPrediction.created_at.desc())
    if device_kind is not None:
        statement = statement.where(MLOfflineRiskPrediction.device_kind == device_kind)
    rows = session.exec(statement.limit(limit)).all()
    return MLOfflineRiskPredictionsPublic(data=rows, count=len(rows))


@router.get("/models/status", response_model=MLModelsStatusPublic)
def read_model_status(
    session: SessionDep,
    current_user: CurrentUser,
    limit: int = Query(default=50, ge=1, le=200),
) -> MLModelsStatusPublic:
    del current_user
    rows = session.exec(select(MLModelRegistry).order_by(MLModelRegistry.trained_at.desc()).limit(limit)).all()
    data = [
        {
            "model_family": row.model_family,
            "version": row.version,
            "status": row.status,
            "train_rows": row.train_rows,
            "metric_primary": row.metric_primary,
            "metric_secondary": row.metric_secondary,
            "trained_at": row.trained_at,
            "activated_at": row.activated_at,
        }
        for row in rows
    ]
    return MLModelsStatusPublic(data=data, count=len(data))


@router.post("/run-cycle", dependencies=[Depends(get_current_active_superuser)], response_model=Message)
async def run_ml_cycle(current_user: CurrentUser) -> Message:
    del current_user
    if not settings.ML_ENABLED:
        raise HTTPException(status_code=503, detail="ML is disabled")
    try:
        async with httpx.AsyncClient(timeout=120) as client:
            resp = await client.post(f"{settings.ML_SERVICE_URL.rstrip('/')}/run-cycle")
        if resp.status_code >= 400:
            raise HTTPException(status_code=502, detail=f"ML service error: {resp.text[:300]}")
        return Message(message="ML cycle started")
    except httpx.HTTPError as exc:
        raise HTTPException(status_code=502, detail=f"ML service unavailable: {exc}") from exc

File: frontend/src/api/ml.ts
import api from "./http";

export interface MLTonerPrediction {
  id: string;
  printer_id: string | null;
  printer_name: string | null;
  toner_color: "black" | "cyan" | "magenta" | "yellow" | string;
  toner_model: string | null;
  current_level: number | null;
  days_to_replacement: number | null;
  predicted_replacement_at: string | null;
  confidence: number;
  model_version: string;
  created_at: string;
}

export interface MLOfflineRiskPrediction {
  id: string;
  device_kind: string;
  device_id: string | null;
  device_name: string | null;
  address: string | null;
  risk_score: number;
  risk_level: "low" | "medium" | "high" | string;
  confidence: number;
  model_version: string;
  created_at: string;
}

export interface MLModelStatus {
  model_family: string;
  version: string;
  status: string;
  train_rows: number;
  metric_primary: number | null;
  metric_secondary: number | null;
  trained_at: string;
  activated_at: string | null;
}

export async function getTonerPredictions(printer_id?: string) {
  const params: Record<string, string> = {};
  if (printer_id) params.printer_id = printer_id;
  const { data } = await api.get<{ data: MLTonerPrediction[]; count: number }>("/ml/predictions/toner", { params });
  return data;
}

export async function getOfflineRiskPredictions(device_kind?: string) {
  const params: Record<string, string> = {};
  if (device_kind) params.device_kind = device_kind;
  const { data } = await api.get<{ data: MLOfflineRiskPrediction[]; count: number }>("/ml/predictions/offline-risk", {
    params,
  });
  return data;
}

export async function getMLModelStatus() {
  const { data } = await api.get<{ data: MLModelStatus[]; count: number }>("/ml/models/status");
  return data;
}

export async function runMLCycle() {
  const { data } = await api.post<{ message: string }>("/ml/run-cycle");
  return data;
}
